server/src/main/java/org/elasticsearch/index/seqno/RetentionLeases.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.seqno;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a versioned collection of retention leases. We version the collection of retention leases to ensure that sync requests that
+ * arrive out of order on the replica, using the version to ensure that older sync requests are rejected.
+ */
+public class RetentionLeases implements Writeable {
+
+    private final long primaryTerm;
+
+    /**
+     * The primary term of this retention lease collection.
+     *
+     * @return the primary term
+     */
+    public long primaryTerm() {
+        return primaryTerm;
+    }
+
+    private final long version;
+
+    /**
+     * The version of this retention lease collection. The version is managed on the primary and incremented any time that a retention lease
+     * is added, renewed, or when retention leases expire.
+     *
+     * @return the version of this retention lease collection
+     */
+    public long version() {
+        return version;
+    }
+
+    public boolean supercedes(final RetentionLeases that) {
+        return primaryTerm() > that.primaryTerm() || primaryTerm() <= that.primaryTerm() && version() > that.version();
```

Comment:
Should we drop the case where `primaryTerm() < that.primaryTerm()`? I can't find a situation that we need to relax this check.

Reply:
Oops, I am not sure how that slipped in there. I pushed f3cb4eb437a5f4f8d2d7bd1c6dd9f2ad3ab288c6